Kernel/Graphics/VirtIOGPU/DisplayConnector.cpp

```diff
@@ -0,0 +1,788 @@
+                return EOVERFLOW;
+            }
+            auto target = transfer_buffer_region.vaddr().offset(transfer_descriptor.offset_in_region).as_ptr();
+            return copy_from_user(target, transfer_descriptor.data, transfer_descriptor.num_bytes);
+        } else if (transfer_descriptor.direction == VIRGL_DATA_DIR_HOST_TO_GUEST) {
+            if (transfer_descriptor.offset_in_region + transfer_descriptor.num_bytes > NUM_TRANSFER_REGION_PAGES * PAGE_SIZE) {
+                return EOVERFLOW;
+            }
+            auto source = transfer_buffer_region.vaddr().offset(transfer_descriptor.offset_in_region).as_ptr();
+            return copy_to_user(transfer_descriptor.data, source, transfer_descriptor.num_bytes);
+        } else {
+            return EINVAL;
+        }
+    }
+    case VIRGL_IOCTL_SUBMIT_CMD: {
+        auto context_id = TRY(get_context_for_description(description))->context_id();
+        SpinlockLocker locker(m_operation_lock);
+        auto user_command_buffer = static_ptr_cast<VirGLCommandBuffer const*>(arg);
+        auto command_buffer = TRY(copy_typed_from_user(user_command_buffer));
+        submit_command_buffer(context_id, [&](Bytes buffer) {
+            auto num_bytes = command_buffer.num_elems * sizeof(u32);
+            VERIFY(num_bytes <= buffer.size());
+            MUST(copy_from_user(buffer.data(), command_buffer.data, num_bytes));
+            return num_bytes;
+        });
+        return {};
+    }
+    case VIRGL_IOCTL_CREATE_RESOURCE: {
+        auto per_context_state = TRY(get_context_for_description(description));
+        auto user_spec = static_ptr_cast<VirGL3DResourceSpec const*>(arg);
+        VirGL3DResourceSpec spec = TRY(copy_typed_from_user(user_spec));
+
+        Graphics::VirtIOGPU::Protocol::Resource3DSpecification const resource_spec = {
+            .target = static_cast<Graphics::VirtIOGPU::Protocol::Gallium::PipeTextureTarget>(spec.target),
+            .format = spec.format,
+            .bind = spec.bind,
+            .width = spec.width,
+            .height = spec.height,
+            .depth = spec.depth,
+            .array_size = spec.array_size,
+            .last_level = spec.last_level,
+            .nr_samples = spec.nr_samples,
+            .flags = spec.flags,
+            .padding = 0,
+        };
+        SpinlockLocker locker(m_operation_lock);
+        auto resource_id = create_3d_resource(resource_spec).value();
+        attach_resource_to_context(resource_id, per_context_state->context_id());
+        ensure_backing_storage(resource_id, per_context_state->transfer_buffer_region(), 0, NUM_TRANSFER_REGION_PAGES * PAGE_SIZE);
+        spec.created_resource_id = resource_id;
+        // FIXME: We should delete the resource we just created if we fail to copy the resource id out
+        return copy_to_user(static_ptr_cast<VirGL3DResourceSpec*>(arg), &spec);
+    }
+    }
+    return DisplayConnector::ioctl(description, request, arg);
```

Comment:
Nit: We probably add some way of preventing arbitrary userspace programs from writing to the framebuffer, if we haven't already.

Reply:
We don't have a way to prevent it currently (in master too), so let's worry about this later on :)